src/arch/arm/64/kernel/vspace.c

```diff
@@ -1621,8 +1663,29 @@ static exception_t decodeARMFrameInvocation(word_t invLabel, word_t length,
             current_syscall_error.type = seL4_IllegalOperation;
             return EXCEPTION_SYSCALL_ERROR;
         }
+#else
+        /* When in EL1, the mapping must be writeable for DC IVAC */
+        if (invLabel == ARMPageInvalidate_Data) {
+            lookupPTSlot_ret_t lu_ret = lookupPTSlot(find_ret.vspace_root, vaddr);
+            pte_t pte = *lu_ret.ptSlot;
+            void *base_ptr = (void *) cap_frame_cap_get_capFBasePtr(cap);
+            /* The cap mapping info could be out of date, so we need to check that the page we are
+               getting is the one the cap provides authority for. */
+            if (unlikely(lu_ret.ptBitsLeft != pageBitsForSize(cap_frame_cap_get_capFSize(cap)) ||
+                         !pte_is_page_type(pte) ||
+                         pte_get_page_base_address(pte) != pptr_to_paddr(base_ptr))) {
```

Comment:
Shouldn't this be check be introduced for all cache operations? Can cache clean operations cause access faults if there is no mapping present in the page table?
EG if a frame was mapped read-only, and then the vspace was recycled so that there was no mapping in the virtual address before calling `ARMPageClean_Data`.

Reply:
You are right, this is a potential new failure mode that we hadn't thought about yet. It also means that a stale frame cap could be used to flush a pretty much randomly different frame that happens to be mapped at the same vaddr in the new VSpace. So this check should be hoisted up and be done for all flush types on page caps.

I'll put up a separate PR for it.